tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -153,122 +201,82 @@ pointToPos (TSPoint line column) = Pos (fromIntegral line) (fromIntegral column)
-push m = do
-  hasChildren <- ask >>= liftIO . ts_tree_cursor_goto_first_child
-  if hasChildren then do
-    a <- m
-    Just a <$ (ask >>= liftIO . ts_tree_cursor_goto_parent)
-  else
-    pure Nothing
-
 -- | Move the cursor to point at the passed 'TSNode'.
-goto :: (Has (Reader (Ptr Cursor)) sig m, MonadIO m) => TSNode -> m ()
-goto node = do
-  cursor <- ask
-  liftIO (with node (ts_tree_cursor_reset_p cursor))
-
--- | Return the 'Node' that the cursor is pointing at.
-peekNode :: (Has (Reader (Ptr Cursor)) sig m, MonadIO m) => m Node
-peekNode = do
-  cursor <- ask
-  liftIO $ alloca $ \ tsNodePtr -> do
-    _ <- ts_tree_cursor_current_node_p cursor tsNodePtr
-    alloca $ \ nodePtr -> do
-      ts_node_poke_p tsNodePtr nodePtr
-      peek nodePtr
-
--- | Return the field name (if any) for the node that the cursor is pointing at (if any), or 'Nothing' otherwise.
-peekFieldName :: (Has (Reader (Ptr Cursor)) sig m, MonadIO m) => m (Maybe FieldName)
-peekFieldName = do
-  cursor <- ask
-  fieldName <- liftIO $ ts_tree_cursor_current_field_name cursor
-  if fieldName == nullPtr then
-    pure Nothing
-  else
-    Just . FieldName . toHaskellCamelCaseIdentifier <$> liftIO (peekCString fieldName)
+goto :: Ptr Cursor -> TSNode -> MatchM ()
+goto cursor node = liftIO (with node (ts_tree_cursor_reset_p cursor))
 
 
-type Fields = Map.Map FieldName [Node]
+type Fields = [(FieldName, Node)]
 
 -- | Return the fields remaining in the current branch, represented as 'Map.Map' of 'FieldName's to their corresponding 'Node's.
-getFields :: (Has (Reader (Ptr Cursor)) sig m, MonadIO m) => m Fields
-getFields = go Map.empty
-  where go fs = do
-          node <- peekNode
-          fieldName <- peekFieldName
-          keepGoing <- step
-          let fs' = case fieldName of
-                Just fieldName' -> Map.insertWith (flip (++)) fieldName' [node] fs
-                -- NB: We currently skip “extra” nodes (i.e. ones occurring in the @extras@ rule), pending a fix to https://github.com/tree-sitter/haskell-tree-sitter/issues/99
-                _ -> if nodeIsNamed node /= 0 && nodeIsExtra node == 0
-                  then Map.insertWith (flip (++)) (FieldName "extraChildren") [node] fs
-                  else fs
-          if keepGoing then go fs'
-          else pure fs'
+getFields :: Ptr Cursor -> Node -> MatchM Fields
+getFields cursor node
+  | maxCount == 0 = pure []
+  | otherwise     = do
+    nodes <- liftIO . allocaArray maxCount $ \ ptr -> do
+      actualCount <- ts_tree_cursor_copy_child_nodes cursor ptr
+      peekArray (fromIntegral actualCount) ptr
+    traverse (\ node -> (, node) <$> getFieldName node) nodes
+  where
+  maxCount = fromIntegral (nodeChildCount node)
+  getFieldName node
+    | nodeFieldName node == nullPtr = pure (FieldName "extraChildren")
+    | otherwise                     = FieldName . toHaskellCamelCaseIdentifier <$> liftIO (peekCString (nodeFieldName node))
```

Comment:
We get a slight speedup from copying all the child nodes across en masse, filtering on the C side, and then filtering the lists for each field.